Add c-ares resolver tests against GCE DNS, redo of #12010

apolcyn opened the pull request:
This is a redo of https://github.com/grpc/grpc/pull/12010, now using the infrastructure added in #12210.

This adds a "background job" that tests c-ares resolver against GCE DNS. It allows testing that c-ares interops with GCE DNS. IMO another benefit is it adds a sanity check that the local testing DNS server from #12210 is realistic.

This is running in an adhoc jenkins right now, in https://grpc-testing.appspot.com/job/gRPC_naming_adhoc/ (that job previously used by #12010).

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                 nows_per_iteration
----------------------------------------------------------------------------------------  --------------------
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/16M/1/0  -5%
```

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                   cpu_time    locks_per_iteration    nows_per_iteration    real_time
--------------------------------------------------------------------------  ----------  ---------------------  --------------------  -----------
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<3, true>>/0/16k     +4%                                                      +4%
BM_PumpStreamClientToServer<SockPair>/128M                                              +9%
BM_StreamingPingPongMsgs<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/16M                                     -7%
```

apolcyn commented:
thanks for the review!

also doing a late cc of @markdroth and @nicolasnoble, to boost awareness of this background test

markdroth commented on templates/test/cpp/naming/resolver_gce_integration_tests_defs.include:
```diff
@@ -0,0 +1,65 @@
+if [[ "$GRPC_DNS_RESOLVER" == "" ]]; then
+  export GRPC_DNS_RESOLVER=ares
+fi
+if [[ "$GRPC_DNS_RESOLVER" != ares ]]; then
```
This could be `elif`.

apolcyn replied:
done

markdroth commented on test/cpp/naming/resolver_gce_integration_tests_runner.sh:
```diff
@@ -0,0 +1,360 @@
+if [[ "$GRPC_DNS_RESOLVER" == "" ]]; then
+  export GRPC_DNS_RESOLVER=ares
+fi
+if [[ "$GRPC_DNS_RESOLVER" != ares ]]; then
```
`elif`

apolcyn replied:
done

markdroth commented on test/cpp/naming/gen_build_yaml.py:
```diff
@@ -33,21 +39,107 @@ def _build_expected_addrs_cmd_arg(expected_addrs):
+    uploadable = r_data[0].split(' ')
+    uploadable[3] = uploadable_target
+    return '\"%s\"' % ' '.join(uploadable)
+  if r_type == 'TXT':
```
It would be really useful to have a stand-alone script that would convert a service config string in JSON form into the form needed for a DNS zone file (which is presumably the same form used in gcloud).  This would be used by folks in the OSS world that want to encode the service config in their DNS server.  Can you refactor this to provide such a utility?

Feel free to do this in a separate PR.  It doesn't need to block this.

apolcyn replied:
SG, I think that gcloud has an alternate mode of uploading records that involves pointing it to a zone file, rather than specifying each record via command line parameters. I might try changing this around to use such a JSON -> TXT script a part of creating a zone file.

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.old: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.old: 1


[microbenchmarks] Performance differences noted:
Benchmark                                                                                    nows_per_iteration
-------------------------------------------------------------------------------------------  --------------------
BM_StreamingPingPongMsgs<InProcessCHTTP2, NoOpMutator, NoOpMutator>/16M                      -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/16M/1/0  -6%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/2M                                     -5%
```

apolcyn commented:
I signed it!

apolcyn commented:
just rebased on lastest master and fixed an issue with the template (previously `templates/test/cpp/naming/resolver_gce_integration_tests_runner.sh.template` used `resolver_component_test_cases` rather than `resolver_gce_integration_test_cases`, which was a noticeable issue after rebasing).

grpc-jenkins commented:
```
[trickle] No significant performance differences
```

grpc-jenkins commented:
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                                       atm_add_per_iteration    cpu_time    locks_per_iteration    nows_per_iteration    real_time
----------------------------------------------------------------------------------------------  -----------------------  ----------  ---------------------  --------------------  -----------
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<1, true>>/0/16k                                                  +8%                                                      +8%
BM_HpackEncoderEncodeHeader<SingleNonInternedBinaryElem<10, false>>/0/16k                                                +6%                                                      +6%
BM_HpackEncoderEncodeHeader<SingleNonInternedElem>/0/16k                                                                 +4%                                                      +4%
BM_HpackParserParseHeader<NonIndexedBinaryElem<1, true>>                                                                 +6%                                                      +6%
BM_IsolatedFilter<ClientDeadlineFilter, NoOp>                                                                            +5%                                                      +6%
BM_PumpStreamServerToClient<InProcess>/512                                                                               +4%                                                      +4%
BM_PumpStreamServerToClient<SockPair>/8                                                                                                                                           +4%
BM_StreamCreateSendInitialMetadataDestroy<RepresentativeClientInitialMetadata>                                           +4%                                                      +4%
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/8/2                                                         +7%                                                      +8%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/1/1/1                                         +5%                                                      +5%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/512/1/0                                                                                                +4%
BM_StreamingPingPongWithCoalescingApi<InProcessCHTTP2, NoOpMutator, NoOpMutator>/2M/2/0                                              -4%
BM_StreamingPingPongWithCoalescingApi<MinInProcess, NoOpMutator, NoOpMutator>/0/0/1                                      +5%                                                      +5%
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<100>, 2>, NoOpMutator>/0/0                           +4%                                                      +4%
BM_UnaryPingPong<InProcess, Client_AddMetadata<RandomBinaryMetadata<10>, 1>, NoOpMutator>/0/0                            +9%                                                      +9%
BM_UnaryPingPong<InProcess, NoOpMutator, NoOpMutator>/256k/256k                                                          +5%                                                      +5%
BM_UnaryPingPong<InProcessCHTTP2, NoOpMutator, NoOpMutator>/4k/0                                                         +6%                                                      +6%
BM_UnaryPingPong<MinInProcess, NoOpMutator, NoOpMutator>/0/4k                                                            +6%                                                      +6%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/64/64                                                     +9%                                                      +9%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/0                                                       +9%                                                      +9%
BM_UnaryPingPong<MinInProcessCHTTP2, NoOpMutator, NoOpMutator>/8/8                                                       +4%                                                      +4%
BM_UnaryPingPong<MinTCP, NoOpMutator, NoOpMutator>/2M/2M                                        -5%                                                         -7%
```

apolcyn commented:
* basic tests linux c++ opt kokoro - https://github.com/grpc/grpc/issues/12313
* interop cloud to prod:
  - https://github.com/grpc/grpc/issues/12261 `php oauth2_auth_token`
  - https://github.com/grpc/grpc/issues/12618
  - https://github.com/grpc/grpc/issues/10763 ( `csharpcoreclr large_unary`)
